File: apps/passenger-android/src/main/kotlin/lk/mageride/passenger/di/PassengerDatabase.kt
package lk.mageride.passenger.di

import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import lk.mageride.shared.db.MageRideDatabaseFactory
import lk.mageride.shared.db.passenger.PassengerDb

/**
 * The one open `mageride_passenger.db`, for the whole process.
 *
 * **C018 deliberately binds no open database** — opening one is `suspend` because the SQLCipher key
 * comes out of the Android Keystore, and Koin has no suspending factory, so a `single { runBlocking
 * { … } }` would put a Keystore round trip on `Application.onCreate`. `localDbModule`'s KDoc says
 * the app should *"open the database once during start-up and bind the result in its own module"*.
 * This is that binding, deferred: the first caller opens it, everybody after that gets the same
 * handle, and an app that never touches the local tables never pays for the key at all.
 *
 * **One handle, not one per caller.** `mobile_db_schema.md` §2 gives this app six tables and every
 * screen group from C077 on will read one of them — saved addresses, place recents, the ride
 * projection, cached estimates, the proxy location requests, the block list. Six `openPassenger()`
 * calls would be six SQLite connections to one encrypted file, each with its own write lock. The
 * [Mutex] is what makes "first caller opens it" true when two of them arrive at once; it guards the
 * *open*, not the queries, which are the driver's own business. This is the shape the C075 handoff
 * asked this component to copy, and the mistake it asked it not to repeat.
 *
 * **Nothing in the shell reads it yet**, and that is deliberate rather than an omission: the shell
 * owns no screen and has no table of its own. It is bound here because DI is this component's
 * deliverable and because the alternative — each of C077–C084 opening the file itself — is exactly
 * the bug above.
 *
 * The handle is never closed. Its lifetime is the process's, and closing it would be closing it
 * under whichever screen happens to be reading.
 */
internal class PassengerDatabase(private val factory: MageRideDatabaseFactory) {

    private val mutex = Mutex()

    @Volatile
    private var opened: PassengerDb? = null

    /**
     * The database, opening it on the first call.
     *
     * **Every query on the result blocks** (`MageRideDb`'s KDoc) — SQLDelight's Android driver is
     * synchronous — so a caller runs them off the main thread with `withContext(Dispatchers.IO)`.
     * A view model must not call this from a composition.
     */
    suspend fun get(): PassengerDb = opened ?: mutex.withLock {
        opened ?: factory.openPassenger().also { opened = it }
    }
}

File: apps/passenger-ios/PassengerApp/DI/PassengerDatabase.swift
import Foundation
import MageRideShared

/// The app's one open `mageride_passenger.db`, opened on first use and shared.
///
/// **Why a holder rather than a property on ``PassengerGraph``.** Opening the database is `suspend`
/// on the Kotlin side — the file is under `NSFileProtectionCompleteUntilFirstUserAuthentication` and
/// `MageRideDatabaseFactory.openPassenger` is an `async` call — so it cannot be a stored property
/// built in an initialiser without putting a round trip in front of the first frame. `localDbModule`
/// binds two things and deliberately **not** the database for exactly this reason; the app owns the
/// handle.
///
/// **One handle, many callers.** `mobile_db_schema.md` §2 gives this app six tables and eight screen
/// groups will read them — C096's `place_recents`, C099's package projections, C101's addresses.
/// Eight groups calling `openPassenger()` independently would be eight SQLite connections to one
/// encrypted file, which is the defect C075 fixed on the driver side and C093 turned into
/// `DriverDatabase`. Do not call `openPassenger()` anywhere else.
///
/// **An `actor`, not a `@MainActor` class.** Two things follow. The `await` inside ``get()`` is a
/// suspension point, so two callers arriving together would both find `database == nil` and both
/// open one — an actor serialises them, which is the same job the `Mutex` does in
/// `apps/passenger-android/.../di/PassengerDatabase.kt`. And every call on the returned handle is
/// **blocking** (SQLDelight's Native driver is synchronous, as `:shared`'s own docs say twice), so
/// the work belongs off the main thread; an actor is where a caller naturally puts it.
///
/// A failed open answers `nil` rather than throwing. There is nothing a passenger can do about a
/// protected file that will not open, and every caller degrades to "no local history" — the recents
/// list is empty, the map still works, and a ride can still be booked.
actor PassengerDatabase {

    private let factory: MageRideDatabaseFactory
    private var database: PassengerDb?

    init(factory: MageRideDatabaseFactory) {
        self.factory = factory
    }

    /// The open database, opening it on the first call.
    ///
    /// Idempotent and race-safe: a second caller during the first open waits on the actor rather
    /// than starting a second `openPassenger`.
    func get() async -> PassengerDb? {
        if let database { return database }
        let opened = try? await factory.openPassenger(inMemory: false)
        database = opened
        return opened
    }
}
